AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt

```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
+        val oggFiles = cacheDir.listFiles { file ->
+            file.parentFile == cacheDir &&
+            file.name.matches(Regex("^audio_-?\\d+\\.ogg$"))
+        } ?: emptyArray()
+        val filesToDelete = oggFiles.filter { file -> !oggCacheMap.values.contains(file) }
+        for (file in filesToDelete) {
+            file.delete()
+        }
+    }
+
+    override fun start(context: Context) {
+        deleteOggCacheFiles(cacheDir = context.cacheDir, oggCacheMap = oggFileCache)
+
+        patcher.after<WidgetChatListAdapterItemAttachment>(
+            "configureFileData",
+            MessageAttachment::class.java,
+            MessageRenderContext::class.java
+        ) {
+            val messageAttachment = it.args[0] as MessageAttachment
+            val root = WidgetChatListAdapterItemAttachment.`access$getBinding$p`(this).root as ConstraintLayout
+            val card = root.findViewById<MaterialCardView>(attachmentCardId)
+            val ctx = root.context
+
+            card.findViewById<MaterialCardView>(playerBarId)?.let { card.removeView(it) }
+            val loadingBarId = playerBarId + 1
+            card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
+
+            if (!isAudioFile(messageAttachment.filename)) return@after
+
+            val loadingBar = ProgressBar(ctx, null, android.R.attr.progressBarStyleHorizontal).apply {
+                id = loadingBarId
+                isIndeterminate = true
+                layoutParams = FrameLayout.LayoutParams(MATCH_PARENT, 6.dp).apply {
+                    gravity = Gravity.BOTTOM
+                }
+            }
+            card.addView(loadingBar)
+
+            Utils.threadPool.execute {
+                val url = messageAttachment.url
+                val isOgg = messageAttachment.filename.lowercase(Locale.ROOT).endsWith(".ogg")
+                var duration: Long = durationCache[url] ?: run {
+                    val dur = if (isOgg) {
+                        OggMetadataFetcher.fetch(url)?.duration?.times(1000)?.toLong() ?: 0L
+                    } else {
+                        try {
+                            MediaMetadataRetriever().use { retriever ->
+                                retriever.setDataSource(url)
+                                retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLong() ?: 0L
+                            }
+                        } catch (e: Throwable) {
+                            0L
+                        }
+                    }
+                    durationCache[url] = dur
+                    dur
+                }
+
+                Utils.mainThread.post {
+                    card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
+
+                    if (duration == -1L) {
+                        Toast.makeText(ctx, "Failed to load audio metadata.", Toast.LENGTH_SHORT).show()
+                        return@post
```

Comment:
Don't return, allow the player to play just keep the duration at 0.
And don't show a toast, it's pointless